confidant/authnz/userauth.py

```diff
@@ -0,0 +1,661 @@
+            'wantAssertionsSigned':
+                app.config['SAML_SECURITY_ASSERTIONS_SIGNED'],
+            'wantNameIdEncrypted': False,
+            "signatureAlgorithm": app.config['SAML_SECURITY_SIG_ALGO'],
+        }
+
+        # Identity provider section
+        idp_data = {
+            'entityId': app.config['SAML_IDP_ENTITY_ID'],
+            'singleSignOnService': {
+                'url': app.config['SAML_IDP_SIGNON_URL'],
+                'binding': 'urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Redirect'
+            },
+        }
+
+        if app.config['SAML_IDP_LOGOUT_URL']:
+            idp_data['singleLogoutService'] = {
+                'url': app.config['SAML_IDP_LOGOUT_URL'],
+                'binding': 'urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Redirect'
+            }
+
+        if app.config['SAML_IDP_CERT_FILE']:
+            idp_data['x509cert'] = self._load_x509_for_saml(
+                app.config['SAML_IDP_CERT_FILE'])
+        if app.config['SAML_IDP_CERT']:
+            idp_data['x509cert'] = app.config['SAML_IDP_CERT']
+
+        # put it all together into the settings
+        data = {
+            'strict': True,  # must not be changed for security
+            'debug': debug,
+            'sp': sp_data,
+            'idp': idp_data,
+            'security': security_data,
+        }
+
+        # if SAML_RAW_JSON_SETTINGS is set, merge the settings in, doing one
+        # level of deep merging.
+        if app.config['SAML_RAW_JSON_SETTINGS']:
+            logging.debug('overriding SAML settings from JSON')
+            dict_deep_update(data, app.config['SAML_RAW_JSON_SETTINGS'])
+
+        logging.debug('Rendered SAML settings: {!r}'.format(data))
+
+        return data
+
+    def log_in(self):
+        """
+        SAML log-in redirect.
+
+        This method initiates the SAML authentication process by directing the
+        browser to forward along an AuthNRequest to the IdP.
+
+        A separate method handles the post-authentication callback, which will
+        hit /v1/saml/consume, processed by consume_saml_assertion().
+        """
+
+        self_page = request.script_root + request.path
+
+        return flask.redirect(self.login_redirect_url(return_to=self_page))
+
+    def consume_saml_assertion(self):
+        """
+        This method is called in routes implementing a SAML attribute consumer
+        service, which receives POST callbacks from the IdP after the user has
+        authenticated.
+        """
+
+        auth = self._saml_auth()
+
+        logging.debug('Processing SAML response')
+
+        try:
+            request_id = session['saml_authn_request_id']
+        except KeyError:
+            logging.warning('No saml_authn_request_id in session')
+            resp = jsonify(errors=['invalid_response'],
+                           message='SAML request failed',
+                           reason=('No AuthNRequest ID from SP found ' +
```

Comment:
No need for the + when the strings are within parens, it'll automatically append them.